src/decoder/lattice-faster-decoder-combine.h

```diff
@@ -0,0 +1,541 @@
+  // the graph of the state corresponding to token t, or the best of
+  // forward-cost[t] if there were no final-probs active on the final frame.
+  // You cannot call this after FinalizeDecoding() has been called; in that
+  // case you should get the answer from class-member variables.
+  void ComputeFinalCosts(unordered_map<Token*, BaseFloat> *final_costs,
+                         BaseFloat *final_relative_cost,
+                         BaseFloat *final_best_cost) const;
+
+  // PruneForwardLinksFinal is a version of PruneForwardLinks that we call
+  // on the final frame.  If there are final tokens active, it uses
+  // the final-probs for pruning, otherwise it treats all tokens as final.
+  void PruneForwardLinksFinal();
+
+  // Prune away any tokens on this frame that have no forward links.
+  // [we don't do this in PruneForwardLinks because it would give us
+  // a problem with dangling pointers].
+  // It's called by PruneActiveTokens if any forward links have been pruned
+  void PruneTokensForFrame(int32 frame_plus_one);
+
+
+  // Go backwards through still-alive tokens, pruning them if the
+  // forward+backward cost is more than lat_beam away from the best path.  It's
+  // possible to prove that this is "correct" in the sense that we won't lose
+  // anything outside of lat_beam, regardless of what happens in the future.
+  // delta controls when it considers a cost to have changed enough to continue
+  // going backward and propagating the change.  larger delta -> will recurse
+  // less far.
+  void PruneActiveTokens(BaseFloat delta);
+
+  /// Processes nonemitting (epsilon) arcs and emitting arcs for one frame
+  /// together. Consider it as a combination of ProcessEmitting() and 
+  /// ProcessNonemitting().
+  void ProcessForFrame(DecodableInterface *decodable);
+
+  /// Processes nonemitting (epsilon) arcs for one frame.
+  /// Called once when all frames were processed or in GetRawLattice().
+  /// Deal With the tokens in map "next_toks_" which would only contains
+  /// emittion tokens from previous frame.
+  /// If you call this function not in the end of an utterance, recover
+  /// should be true.
+  void ProcessNonemitting(bool recover);
+
+  /// The "cur_toks_" and "next_toks_" actually allow us to maintain current
+  /// and next frames. They are indexed by StateId. It is indexed by frame-index
+  /// plus one, where the frame-index is zero-based, as used in decodable object.
+  /// That is, the emitting probs of frame t are accounted for in tokens at
+  /// toks_[t+1].  The zeroth frame is for nonemitting transition at the start of
+  /// the graph.
+  StateIdToTokenMap cur_toks_;
+  StateIdToTokenMap next_toks_;
+
+  /// When we call GetRawLattice() in the middle of an utterance, we have to
+  /// process non-emitting arcs so that we need to recover it original status.
+  std::unordered_map<Token*, BaseFloat> recover_map_;  // Token pointer to tot_cost
```

Comment:
I don't understand your comments about this at all.
I also don't see where you clear this map.  That looks like a bug to me; remember, memory can be reused.